server/src/endpoints/qs/push_notification_provider.rs

```diff
@@ -40,42 +85,125 @@ pub struct ApnsState {
 
 #[derive(Debug, Clone)]
 pub struct ProductionPushNotificationProvider {
+    fcm_state: Option<FcmState>,
     apns_state: Option<ApnsState>,
 }
 
 impl ProductionPushNotificationProvider {
-    // Create a new ProductionPushNotificationProvider. If the config_option is
-    // None, the provider will effectively not send push notifications.
-    pub fn new(config_option: Option<ApnsSettings>) -> Result<Self, Box<dyn std::error::Error>> {
-        let Some(config) = config_option else {
-            return Ok(Self { apns_state: None });
+    // Create a new ProductionPushNotificationProvider. If the settings are
+    // None, the provider will effectively not send push notifications for that
+    // platform.
+    pub fn new(
+        fcm_settings: Option<FcmSettings>,
+        apns_settings: Option<ApnsSettings>,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        // Read the FCN service account file
+        let fcm_state = if let Some(fcm_settings) = fcm_settings {
+            let mut service_account_file = File::open(fcm_settings.path)?;
+            let mut service_account = String::new();
+            service_account_file.read_to_string(&mut service_account)?;
+
+            Some(FcmState {
+                service_account: serde_json::from_str(&service_account)?,
```

Comment:
I would propose to introduce a strong type for the service account key:

```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceAccountKey {
    #[serde(rename = "type")]
    pub key_type: Option<String>,
    pub project_id: Option<String>,
    pub private_key_id: Option<String>,
    pub private_key: Zeroizing<String>,
    pub client_email: String,
    pub client_id: Option<String>,
    pub auth_uri: Option<String>,
    pub token_uri: String,
    pub auth_provider_x509_cert_url: Option<String>,
    pub client_x509_cert_url: Option<String>,
}
```

This is derived from a real account key data. There is a protobuf spec, which for some reason does not entirely fit the real key: https://github.com/googleapis/googleapis/blob/master/google/iam/admin/v1/iam.proto#L822.

Reply:
Thanks for the template! The only field missing was `universe_domain`.